Add more NRT examples (#1225)
dotnet/csharpstandard: standard/types.md

> *Example*: Each of the following expressions include some form of a null check. The null-state of `o` can change from not null to maybe null after each of these statements:
>
> <!-- Example: {template:"code-in-class-lib", name:"NullChecks", expectedWarnings:["CS8602", "CS8602", "CS8602"]} -->
> ```csharp
> #nullable enable
> public void M(string s)
> {
>     int length = s.Length; // No warning. s is not null
>
>     _ = s == null; // Null check by testing equality. The null state of s is maybe null
>     length = s.Length; // Warning, and changes the null state of s to not null
>
>     _ = s?.Length; // The ?. is a null check and changes the null state of s to maybe null
>     if (s.Length > 4) // Warning. Changes null state of s to not null
>     {
>         _ = s?[4]; // ?[] is a null check and changes the null state of s to maybe null
>         _ = s.Length; // Warning. s is maybe null
>     }
> }
> ```

> Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.

Reviewer: ```suggestion
Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state analysis may assume that assignment to the event or property is an assignment to a compiler generated backing field.
```

Reviewer: (Or change "assume that" to "treat".)

> Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state analysis may infer that assignment to the event or property is an assignment to a compiler generated backing field.

Reviewer 2: Uber nitpick :D
```suggestion
Both auto-property and field-like event declarations make use of a compiler-generated backing field. Null state analysis may infer that assignment to the event or property is an assignment to a compiler generated backing field.
```

Reviewer 2: Also "auto property" instead of "auto-property" but that one is likely a separate item if it's even palatable to consider.

Later in the method, the code checks that `s` is not a null reference. The null-state of `s` can change to maybe null after the null-checked block closes. A compiler can infer that `s` is maybe null because the code was written to assume that it might have been null. Generally, when the code contains a null check, a compiler may infer that the value might have been null:

> *Example*: Each of the following expressions include some form of a null check. The null-state of `o` can change from not null to maybe null after each of these statements:
>
> <!-- Example: {template:"code-in-class-lib", name:"NullChecks", expectedWarnings:["CS8600","CS8602", "CS8602", "CS8602"]} -->
> ```csharp
> #nullable enable
> public void M(string s)
> {
>     // s is not null:
>     int length = s.Length;
>

> >     _ = s = null; // Null check by testing equality. The null state of o is maybe null

Reviewer: Is this meant to be `s == null`? At the moment this is an assignment...

Reviewer: (And what is "o" referred to in the comment? I'm guessing s used to be o - needs to be changed below in multiple places.)

> >     _ = s = null; // Null check by testing equality. The null state of o is maybe null

Author: ```suggestion
>     _ = s == null; // Null check by testing equality. The null state of s is maybe null
```

> >     length = s.Length; // warning, and changes the null state of o to not null

Reviewer: ```suggestion
>     length = s.Length; // Warning, and changes the null state of o to not null
```

>
>     _ = s?.Length; // The ?. is a null check and changes the null state of o to maybe null
>     if (s.Length > 4) // Warning. Changes null state of o to not null
>     {

> >        _ = s?[4]; //is a null check and changes the null state of o to maybe null

Reviewer: We're missing a noun at the start of the sentence, and a space after `//`

>         _ = s.Length; // warning
>     }
> }
> ```

> Both auto-property and field like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.

Reviewer: ```suggestion
Both auto-property and field-like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to a compiler generated backing field.
```

Reviewer: (I'd personally prefer to hyphenate compiler-generated everywhere as well, but I'm not doing that now. Field-like events are hyphenated consistently though, I believe.)

> > *Example*: A compiler can determine that writing an auto-property or field like event writes the corresponding compiler generated backing field. The null state of the property matches that of the backing field.

Reviewer: ```suggestion
> *Example*: A compiler can determine that writing an auto-property or field-like event writes the corresponding compiler generated backing field. The null state of the property matches that of the backing field.
```

>
> <!-- Example: {template:"standalone-console", name:"NullPropertyAnalysis", expectedException:"NullReferenceException"} -->
> ```csharp
> class Test
> {
>     public string P
>     {
>         get;
>         set;
>     }
>
>     public Test() {} // Warning. "P" not set to a non-null value.
>
>     static void Main()
>     {
>         var t = new Test();

> >         int len = t.P.Length; // No warning. null state is not null

Reviewer: ```suggestion
>         int len = t.P.Length; // No warning. Null state is not null.
```

>     }
> }
> ```
>

> > In the previous example, the constructor doesn't set `P` to a not null value, and a compiler issues a warning. There's no warning when the `P` property is accessed, because the type of the property is a non nullable reference type. *end example*

Reviewer: > and a compiler issues a warning

Issues, or may issue?

> *Example*: Each of the following expressions include some form of a null check. The null-state of `o` can change from not null to maybe null after each of these statements:
>
> <!-- Example: {template:"code-in-class-lib", name:"NullChecks", expectedWarnings:["CS8600","CS8602", "CS8602", "CS8602"]} -->
> ```csharp
> #nullable enable
> public void M(string s)
> {
>     // s is not null:
>     int length = s.Length;
>
>     _ = s = null; // Null check by testing equality. The null state of o is maybe null
>     length = s.Length; // warning, and changes the null state of o to not null
>
>     _ = s?.Length; // The ?. is a null check and changes the null state of o to maybe null
>     if (s.Length > 4) // Warning. Changes null state of o to not null
>     {
>        _ = s?[4]; //is a null check and changes the null state of o to maybe null
>         _ = s.Length; // warning
>     }
> }

> > ```

Reviewer 3: Should the next line (964) markdownlint-disable be deleted? It isn't paired with an enable.

<!-- markdownlint-disable MD028 -->

Both auto-property and field like event declarations make use of a compiler generated backing field. Null state may assume that assignment to the event or property as assignment to the compiler generated backing field.

> *Example*: A compiler can determine that writing an auto-property or field like event writes the corresponding compiler generated backing field. The null state of the property matches that of the backing field.
>
> <!-- Example: {template:"standalone-console", name:"NullPropertyAnalysis", expectedException:"NullReferenceException"} -->
> ```csharp
> class Test
> {
>     public string P
>     {
>         get;
>         set;
>     }
>
>     public Test() {} // Warning. "P" not set to a non-null value.
>
>     static void Main()
>     {
>         var t = new Test();
>         int len = t.P.Length; // No warning. null state is not null
>     }
> }
> ```
>

> > In the previous example, the backing field for the `DisappearingProperty` is set to null when it is read. However, a compiler may assume that reading a property doesn’t change the null state of that expression. *end example*

Reviewer 3: Example above has been replaced, this para describes the old version.

>         var t = new Test();
>         if (t.DisappearingProperty != null)
>         {

> >             int len = t.DisappearingProperty.Length; // No warning. A compiler can assume state

Reviewer: I suspect the second sentence *is* as intended, but sounds incomplete to me.

Author: reworded that sentence.

>         }
>     }
> }
> ```
>
> In the previous example, the backing field for the `DisappearingProperty` is set to null when it is read. However, a compiler may assume that reading a property doesn’t change the null state of that expression. *end example*

A compiler may use any expression that dereferences a variable, property, or event to set the null state to not null. If it were null, the dereference expression would have thrown a `NullReferenceException`:

> *Example*:
>
> <!-- Example: {template:"standalone-lib-without-using", name:"ChainedAccess", ignoredWarnings:["CS0649", "CS8602"]} -->
> ```csharp
>
> public class C
> {
>     private C? child;
>    
>     public void M()
>     {
>         _ = child.child.child; // Warning. Dereference possible null value
>         var greatGrandChild = child.child.child; // No warning. 

> >

Reviewer: Remove blank line?

>
> In the previous example, the backing field for the `DisappearingProperty` is set to null when it is read. However, a compiler may assume that reading a property doesn’t change the null state of that expression. *end example*

> A compiler may use any expression that dereferences a variable, property, or event to set the null state to not null. If it were null, the dereference expression would have thrown a `NullReferenceException`:

Reviewer: Do we want to add a note about extension methods, where calling an extension method *looks* like it dereferences a variable, but doesn't actually do so?
